dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/EntityToDbTableMapping.java

```diff
@@ -0,0 +1,177 @@
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.cacheinvalidation;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+import javax.persistence.metamodel.EntityType;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.hibernate.SessionFactory;
+import org.hibernate.metamodel.internal.MetamodelImpl;
+import org.hibernate.persister.collection.BasicCollectionPersister;
+import org.hibernate.persister.collection.CollectionPersister;
+import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.persister.entity.SingleTableEntityPersister;
+import org.hibernate.type.CollectionType;
+import org.hibernate.type.Type;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Slf4j
+@Component
+public class EntityToDbTableMapping
+{
+    @Autowired
+    private SessionFactory sessionFactory;
+
+    private final TreeMap<String, List<Object[]>> entityToTableNames = new TreeMap<>();
+
+    protected void init()
+    {
+        extractTableNamesFromHibernateMetamodel();
+
+        log.debug( "Finished extracting table names from the Hibernate metamodel. Entity table=\n" + printEntityTable(
+            entityToTableNames ) );
+    }
+
+    protected List<Object[]> get( String tableName )
+    {
+        if ( entityToTableNames.size() == 0 )
+        {
+            throw new IllegalStateException( "EntityToDbTableMapping is not initialized yet!" );
+        }
+        return entityToTableNames.get( tableName );
+    }
+
+    private void extractTableNamesFromHibernateMetamodel()
+    {
+        for ( EntityType<?> entity : sessionFactory.getMetamodel().getEntities() )
+        {
+            extractTableNames( entity.getJavaType() );
+        }
+    }
+
+    public void extractTableNames( final Class<?> modelClazz )
+    {
+        final MetamodelImpl metamodel = (MetamodelImpl) sessionFactory.getMetamodel();
```

Comment:
Can you check if you can use [schema.getTableName() ](https://github.com/dhis2/dhis2-core/blob/c619fe547ad791b0bc8b940ac5a7b8bd2538246a/dhis-2/dhis-services/dhis-service-schema/src/main/java/org/hisp/dhis/schema/Schema.java#L550)

Reply:
I would rather prefer to use Hibernate's metadata model here, since this is strictly working with the database layer. I also need the more complex relationships like many-to-many collections from the model.